[Bug]: Build fails with LTO

### Expected behavior?

Build should succeed.

### Actual behavior?

Build fails.

### Steps to reproduce the problem

I tried to build with the following *FLAGS to optimize the build: `-flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing`

Note the -Werror=* flags are used to help detect cases where the compiler tries to optimize by assuming UB cannot exist in the source code -- if it does exist, ordinarily the code would be miscompiled, and this says to make the miscompilation a fatal error.

### Error messages?

I got this error:
```
FAILED: vbam 
: && /usr/bin/x86_64-pc-linux-gnu-g++ -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-clash-protection -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing  -Wformat -Werror=format-security -Wl,-O1 -Wl,--as-needed -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing -Wl,--defsym=__gentoo_check_ldflags__=0 CMakeFiles/vbam.dir/src/sdl/ConfigManager.cpp.o CMakeFiles/vbam.dir/src/sdl/SDL.cpp.o CMakeFiles/vbam.dir/src/sdl/filters.cpp.o CMakeFiles/vbam.dir/src/sdl/text.cpp.o CMakeFiles/vbam.dir/src/sdl/inputSDL.cpp.o CMakeFiles/vbam.dir/src/sdl/expr.cpp.o CMakeFiles/vbam.dir/src/sdl/exprNode.cpp.o CMakeFiles/vbam.dir/src/sdl/expr-lex.cpp.o CMakeFiles/vbam.dir/src/common/version.cpp.o -o vbam  libvbamcore.a  fex/libfex.a  /usr/lib64/libSDL2main.a  /usr/lib64/libSDL2.so  /usr/lib64/libOpenGL.so  /usr/lib64/libGLX.so  /usr/lib64/libGLU.so  /usr/lib64/libz.so  -pipe -Wno-unused-command-line-argument -Wformat -Wformat-security -feliminate-unused-debug-types -march=core2 -fPIC -fopenmp -mfpmath=sse -msse2 -Ofast -fomit-frame-pointer -Wno-error  -Wl,-allow-multiple-definition && :
expr-lex.cpp:187:8: error: type ‘struct yy_buffer_state’ violates the C++ One Definition Rule [-Werror=odr]
debugger-expr-lex.cpp:444:8: note: a different type is defined in another translation unit
expr-lex.cpp:196:15: note: the first difference of corresponding definitions is field ‘yy_buf_size’
debugger-expr-lex.cpp:454:6: note: a field of same name but different type is defined in another translation unit
expr-lex.cpp:187:8: note: type ‘yy_size_t’ should match type ‘int’
lto1: some warnings being treated as errors
lto-wrapper: fatal error: /usr/bin/x86_64-pc-linux-gnu-g++ returned 1 exit status
compilation terminated.
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

Also reported downstream at https://bugs.gentoo.org/926080
Full build log: [build.log](https://github.com/visualboyadvance-m/visualboyadvance-m/files/14716648/build.log)


### What game were you playing

_No response_

### What settings are you using?

I wasn't using the program, just building it.

### Specifications and Build details

VBA-M: 2.1.7 -- but the files with the errors haven't changed since then
OS: Linux
If using Linux, specify if using xorg or Wayland: GCC doesn't care whether it is run under xorg or wayland
CPU: not sure it matters
GPU: not sure it matters
RAM: not sure it matters



### Copyright Compliance

- [X] There are no infringing files attached to this bug report.

> Don't use those lto flags, we set our own using the cmake option ENABLE_LTO

Okay. I won't use those LTO flags.

```
# These are not LTO flags
CXXFLAGS+="-Werror=odr"
LDFLAGS+="-Werror=odr"

# THese are the official LTO flags for vbam
mycmakeargs+=('-DENABLE_LTO=ON')

build_with_cmake "${mycmakeargs[@]}"
```

Same error.

```
FAILED: vbam 
: && /usr/bin/x86_64-pc-linux-gnu-g++ -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-clash-protection   -Wformat -Werror=format-security -Werror=odr -Wl,-O1 -Wl,--as-needed  -Wl,--defsym=__gentoo_check_ldflags__=0 CMakeFiles/vbam.dir/src/sdl/ConfigManager.cpp.o CMakeFiles/vbam.dir/src/sdl/SDL.cpp.o CMakeFiles/vbam.dir/src/sdl/filters.cpp.o CMakeFiles/vbam.dir/src/sdl/text.cpp.o CMakeFiles/vbam.dir/src/sdl/inputSDL.cpp.o CMakeFiles/vbam.dir/src/sdl/expr.cpp.o CMakeFiles/vbam.dir/src/sdl/exprNode.cpp.o CMakeFiles/vbam.dir/src/sdl/expr-lex.cpp.o CMakeFiles/vbam.dir/src/common/version.cpp.o -o vbam  libvbamcore.a  fex/libfex.a  /usr/lib64/libSDL2main.a  /usr/lib64/libSDL2.so  /usr/lib64/libOpenGL.so  /usr/lib64/libGLX.so  /usr/lib64/libGLU.so  /usr/lib64/libz.so  -pipe -Wno-unused-command-line-argument -Wformat -Wformat-security -feliminate-unused-debug-types -march=core2 -fPIC -fopenmp -mfpmath=sse -msse2 -Ofast -fomit-frame-pointer -flto=8 -ffat-lto-objects -Wno-error  -Wl,-allow-multiple-definition && :
expr-lex.cpp:187:8: error: type ‘struct yy_buffer_state’ violates the C++ One Definition Rule [-Werror=odr]
debugger-expr-lex.cpp:444:8: note: a different type is defined in another translation unit
expr-lex.cpp:196:15: note: the first difference of corresponding definitions is field ‘yy_buf_size’
debugger-expr-lex.cpp:454:6: note: a field of same name but different type is defined in another translation unit
expr-lex.cpp:187:8: note: type ‘yy_size_t’ should match type ‘int’
lto1: some warnings being treated as errors
lto-wrapper: fatal error: /usr/bin/x86_64-pc-linux-gnu-g++ returned 1 exit status
compilation terminated.
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

This is because your cmake option ENABLE_LTO simply tells cmake to use its builtin LTO handling, which passes `-flto=8 -ffat-lto-objects`. There's no difference from a code correctness perspective. Well, fat LTO is different in the sense that it is slow and in this case purposeless.

I am not sure what point you're trying to make. Are you saying the vbam code *doesn't* violate the C++ One Definition Rule?

The warning flags I used at the same time as the LTO flag do not change how LTO works, or cause cmake to build a different binary. They just instruct GCC to detect *miscompilations*.

Please also note the ODR violation is in `yy_buffer_state`, across two files:
- expr-lex.cpp
- debugger-expr-lex.cpp

They are both lex-generated files, generating *different* code but under the same name. The ideal solution here is probably to ensure that the generated definitions are named something different. Consider regenerating with the `--prefix` option to flex, to ensure that they don't clash with each other.

The default prefix is always "yy" which is inconvenient whenever it turns out you use lex for multiple files in the same project, but that is POSIX for you.

--- a/cmake/Toolchain-gcc-clang.cmake
+++ b/cmake/Toolchain-gcc-clang.cmake
@@ -2,12 +2,14 @@ if(X86_32 OR X86_64)
     add_compile_options(-mfpmath=sse -msse2)
 endif()
 
-if(X86_64)
-    # Require and optimize for Core2 level support, tune for generic.
-    add_compile_options(-march=core2 -mtune=generic)
-elseif(X86_32)
-    # Optimize for pentium-mmx and tune for generic for older builds.
-    add_compile_options(-march=pentium-mmx -mtune=generic)
+if(UPSTREAM_RELEASE)
+    if(X86_64)
+        # Require and optimize for Core2 level support, tune for generic.
+        add_compile_options(-march=core2 -mtune=generic)
+    elseif(X86_32)
+        # Optimize for pentium-mmx and tune for generic for older builds.
+        add_compile_options(-march=pentium-mmx -mtune=generic)
+    endif()
 endif()
 
 # Common flags.
@@ -54,5 +56,9 @@ if(CMAKE_SYSTEM_NAME STREQUAL FreeBSD)
 endif()
 
 if(VBAM_STATIC)
-    add_link_options("-static-libgcc" "-static-libstdc++" "-Wl,-Bstatic" "-lstdc++" "-lpthread")
+    add_link_options(-static-libgcc -static-libstdc++ -Wl,-Bstatic -lstdc++ -lpthread)
 endif()
+
+# To support LTO, this must always fail.
+add_compile_options(-Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing)
+add_link_options(   -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing)

--- a/src/sdl/expr-lex.cpp
+++ b/src/sdl/expr-lex.cpp
@@ -151,7 +151,7 @@ typedef unsigned int flex_uint32_t;
 
 #ifndef YY_TYPEDEF_YY_BUFFER_STATE
 #define YY_TYPEDEF_YY_BUFFER_STATE
-typedef struct yy_buffer_state* YY_BUFFER_STATE;
+typedef struct yy_buffer_state_sdl* YY_BUFFER_STATE;
 #endif
 
 extern int yyleng;
@@ -184,7 +184,7 @@ typedef size_t yy_size_t;
 
 #ifndef YY_STRUCT_YY_BUFFER_STATE
 #define YY_STRUCT_YY_BUFFER_STATE
-struct yy_buffer_state {
+struct yy_buffer_state_sdl {
     FILE* yy_input_file;
 
     char* yy_ch_buf; /* input buffer */
@@ -1218,7 +1218,7 @@ YY_BUFFER_STATE yy_create_buffer(FILE* file, int size)
 {
     YY_BUFFER_STATE b;
 
-    b = (YY_BUFFER_STATE)yyalloc(sizeof(struct yy_buffer_state));
+    b = (YY_BUFFER_STATE)yyalloc(sizeof(struct yy_buffer_state_sdl));
     if (!b)
         YY_FATAL_ERROR("out of dynamic memory in yy_create_buffer()");
 
@@ -1381,11 +1381,11 @@ static void yyensure_buffer_stack(void)
 		 * immediate realloc on the next call.
          */
         num_to_alloc = 1;
-        (yy_buffer_stack) = (struct yy_buffer_state**)yyalloc(num_to_alloc * sizeof(struct yy_buffer_state*));
+        (yy_buffer_stack) = (struct yy_buffer_state_sdl**)yyalloc(num_to_alloc * sizeof(struct yy_buffer_state_sdl*));
         if (!(yy_buffer_stack))
             YY_FATAL_ERROR("out of dynamic memory in yyensure_buffer_stack()");
 
-        memset((yy_buffer_stack), 0, num_to_alloc * sizeof(struct yy_buffer_state*));
+        memset((yy_buffer_stack), 0, num_to_alloc * sizeof(struct yy_buffer_state_sdl*));
 
         (yy_buffer_stack_max) = num_to_alloc;
         (yy_buffer_stack_top) = 0;
@@ -1398,13 +1398,13 @@ static void yyensure_buffer_stack(void)
         int grow_size = 8 /* arbitrary grow size */;
 
         num_to_alloc = (yy_buffer_stack_max) + grow_size;
-        (yy_buffer_stack) = (struct yy_buffer_state**)yyrealloc((yy_buffer_stack),
-            num_to_alloc * sizeof(struct yy_buffer_state*));
+        (yy_buffer_stack) = (struct yy_buffer_state_sdl**)yyrealloc((yy_buffer_stack),
+            num_to_alloc * sizeof(struct yy_buffer_state_sdl*));
         if (!(yy_buffer_stack))
             YY_FATAL_ERROR("out of dynamic memory in yyensure_buffer_stack()");
 
         /* zero only the new slots.*/
-        memset((yy_buffer_stack) + (yy_buffer_stack_max), 0, grow_size * sizeof(struct yy_buffer_state*));
+        memset((yy_buffer_stack) + (yy_buffer_stack_max), 0, grow_size * sizeof(struct yy_buffer_state_sdl*));
         (yy_buffer_stack_max) = num_to_alloc;
     }
 }
@@ -1423,7 +1423,7 @@ YY_BUFFER_STATE yy_scan_buffer(char* base, yy_size_t size)
         /* They forgot to leave room for the EOB's. */
         return 0;
 
-    b = (YY_BUFFER_STATE)yyalloc(sizeof(struct yy_buffer_state));
+    b = (YY_BUFFER_STATE)yyalloc(sizeof(struct yy_buffer_state_sdl));
     if (!b)
         YY_FATAL_ERROR("out of dynamic memory in yy_scan_buffer()");
 

